impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,184 @@
+        queryString.put("grant_type", "client_credentials");
+        queryString.put("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer");
+        queryString.put("client_assertion", signedJwt);
+        queryString.put("scope", scope);
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.add("Accept", "application/json");
+        httpHeaders.add("Content-Type", "application/x-www-form-urlencoded");
+
+        Request accessTokenRequest = new DefaultRequest(
+            HttpMethod.POST,
+            clientConfiguration.getBaseUrl() + "/oauth2/v1/token",
+            queryString,
+            httpHeaders,
+            new ByteArrayInputStream("".getBytes("UTF-8")),
+            0);
+
+        RequestExecutor requestExecutor = new OkHttpRequestExecutorFactory().create(clientConfiguration);
+
+        Response accessTokenResponse = requestExecutor.executeRequest(accessTokenRequest);
+
+        if (accessTokenResponse != null && accessTokenResponse.getBody() != null) {
+            Reader reader = new InputStreamReader(accessTokenResponse.getBody(), StandardCharsets.UTF_8);
+
+            if (accessTokenResponse.getHttpStatus() == HttpURLConnection.HTTP_OK) {
+                log.debug("OAuth2 access token request was successful");
+                return objectMapper.readValue(reader, OAuth2AccessToken.class);
+            } else {
+                log.debug("OAuth2 access token request failed with HTTP {} error", accessTokenResponse.getHttpStatus());
+
+                OAuth2ErrorResponse errorResponse = objectMapper.readValue(reader, OAuth2ErrorResponse.class);
+
+                if (!Strings.isEmpty(errorResponse.getError()) ||
+                    !Strings.isEmpty(errorResponse.getErrorDescription())) {
+                    throw new HttpException("Received HTTP " + accessTokenResponse.getHttpStatus() +
+                        " response from Authorization Server. error - " + errorResponse.getError() +
+                        ", error description - " + errorResponse.getErrorDescription());
+                } else {
+                    throw new HttpException("Received HTTP " + accessTokenResponse.getHttpStatus() +
+                        " response from Authorization Server");
+                }
+            }
+        } else {
+            throw new HttpException("Could not retrieve access token from Authorization Server");
+        }
+    }
+
+    /**
+     * Create signed JWT string with the supplied {@link ClientConfiguration} details.
+     *
+     * @param {@link ClientConfiguration} object
+     * @return signed JWT string
+     * @throws {@link InvalidKeyException}
+     * @throws {@link IOException}
+     */
+    private String createSignedJWT(ClientConfiguration clientConfiguration)
+        throws InvalidKeyException, IOException {
+        String clientId = clientConfiguration.getClientId();
+        String privateKeyFilePath = clientConfiguration.getPrivateKey();
```

Comment:
probably don't want to parse the key from disk every time it is needed, though assuming the key is good for an hour that might not be an issue and would make for easy key rotation 🤔

Reply:
That would be nice actually. Will consider this improvement/optimization in the future so we dont hold this PR for long?